### ISSUE TITLE
Basic operations on mutable vs immutable collections

### ISSUE THREAD

**Issue by szeiger** [scala/collection-strawman#33]
While working on a blog post about the new collections design I came up with this snippet and realized that we don't have a better solution so far:

```scala
var s1 = collection.immutable.Vector[Int](1)
s1 ++= Seq(2)
val s2 = collection.mutable.ArrayBuffer[Int](1)
s2 ++= Seq(2)
```

The first one is a compound assignment operator, the second one a real `++=` operator. Keeping inheritance hierarchies separate and having separate implementations doesn't help because the two cases are already different and only appear the same syntactically.

Should we try to make these calls look different from each other (which creates additional migration effort)?

How could it be done? One option is to remove all methods ending in `=` from mutable collections on the grounds that if an operation is not a reassignment to a variable it shouldn't look like one. What should they be called instead? `++` is not an option if mutable and immutable collections share common supertypes. We'd have to keep mutable collections completely separate and require that users get a quasi-immutable view even for basic `Iterable` operations.

**Comment by julienrf** [comment]
To me this is not a big deal and illustrates two different usages: a var with an immutable collection or a val with a mutable one.

In my opinion, what can be confusing in the current collections is the fact that mutable collections inherit from operations like `:+` or `+:` (and also `++`, actually) that could be confusing if people write `xs :+ 1` where they intended to write `xs += 1` (and then they would not understand why their collection is not modified).

BTW, I started writing a blog post about the collections too, we should probably merge our blog posts into a single one, what do you think? I pushed my draft here: https://github.com/scala/scala-lang/compare/blogpost-collections

**Comment by Ichoran** [comment]
Weird.  I thought I already answered this one?

Anyway, why not just drop `++=` entirely?  If we don't have any recursive `A =:= Seq[A]` stuff going on, `+=` can be overloaded with either a single element or a collection type.  Then `++=` can be deprecated on mutable collections, and all is well.

**Comment by Ichoran** [comment]
Incidentally, I think Stefan's blog draft is at this point a better first announcement because of how it frames the problem and gives a high-level overview, etc..  On the other hand, do we need just one?  We can have a higher level, "Hey, collections rewrite in progress!" blog, plus an update-ever-so-often with "we're working on this now and this is what is working well and this is what we're still figuring out".

**Comment by odersky** [comment]
This is an area where I emphatically disagree. Notation should bring out hidden connections between equivalent things. And for a program it does in many cases make no difference whether I have

    val s: immutable.Seq[T]
    s ++= xs

or

    val s: mutable.Seq[T]
    s ++= x

The only difference is at what level the mutability applies. So this means that going from one to the other is a very simple refactoring, which it should be. In fact I have often done that refactoring to my own code. 

So in summary: The sameness of the operator always has been fully intentional and is useful IMO. I would hate to see it go away.

**Comment by szeiger** [comment]
My blog post draft is here: https://github.com/typesafehub/scala-team/pull/44. I'm not sure if everyone here has access. I just finished writing this last night before our team meeting so I put it there for review. Since the goal is to publish it on scala-lang.org, should I just open a PR there?

I agree with @Ichoran that my draft provides more context and that we should have more than one blog post. I propose that we start with my post and follow up with one from @julienrf that covers additional topics or interesting details like in the "Simplified class hierarchy" and "In place transformations for mutable collections" sections.

**Comment by julienrf** [comment]
@szeiger Open a PR on scala/scala-lang.

**Comment by szeiger** [comment]
PR: https://github.com/scala/scala-lang/pull/600

**Comment by julienrf** [comment]
I’m closing this issue since I don’t see anything really actionable, feel free to re-open it.
